docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1063 @@
+
+See more examples in the [air-gapped environment section](#air-gapped-environment)
+
+```yaml
+---
+apiVersion: management.cluster.x-k8s.io/v1alpha1
+kind: InfrastructureProvider
+metadata:
+ name: vsphere
+ namespace: capv-system
+spec:
+ version: v0.4.9
+ secretName: vsphere-variables
+
+```
+
+#### Operator Behaviors
+
+##### Installing a provider
+
+In order to install a new Cluster API provider with the management cluster
+operator you have to create a provider as shown above. See the first example
+API usage to create the secret with variables and the provider itself.
+
+When processing a Provider object the operator will apply the following rules.
+
+- Providers with  `spec.Type == CoreProvider` will be installed first; the
+  other providers will be requeued until the core provider exists.
+- Before installing any provider following preflight checks will be executed :
+    - There should not be another instance of the same provider (same Kind, same
+      name) in any namespace.
+    - The Cluster API contract the provider is abiding by, e.g. v1alpha4, must
+      match the contract of the core provider.
+- The operator will set conditions on the Provider object to surface any
+  installation issues such as pre-flight checks and/or order of installation
+  to accurately inform the user.
+- Since the FetchConfiguration is empty and not defined, the operator will
+  apply the embedded fetch configuration for the given kind and
+  `ObjectMeta.Name`. In this case, the operator will fetch artifacts from
+  https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases.
+
+The installation process managed by the operator is consistent with the
+implementation underlying the `clusterctl init` command and includes the
+following steps:
+- Fetching the provider artifacts (the components yaml and the metadata.yaml
+  file)
+- Applying image overrides, if any
+- Replacing variables
+- Applying the resulting yaml to the cluster
+
+As a final consideration, please note that
+- The operator executes installation for 1 provider at time, while `clusterctl
+  init` manages installation of a group of providers with a single operation.
+- `clusterctl init` uses environment variables and a local configuration file,
+  while the operator uses a Secret; given that we want the users to preserve
+  current behaviour in clusterctl, the init operation should be modified to
+  transfer local configuration to the cluster.
+  As part of `clusterctl init`, it will obtain the list of variables required
+  by the provider components and read the corresponding values from the config
+  or environment variables and build the secret.
+  Any image overrides defined in the clusterctl config will also be applied to
+  the provider's components.
+
+
+##### Upgrading a provider
+
+In order to trigger an upgrade of a new Cluster API provider you have to
+change the `spec.Version` field.
+
+Upgrading a provider in the management cluster must abide by the golden rule
```

Comment:
is there a specific order in which providers should be upgraded?

Reply:
We will be following the same ruleset that clusterctl follows. 

I'm assuming that the CoreProvider must go first. We have that rule specified above as part of the install process. That is, when setting up the management cluster, the CoreProvider must be installed first and the other providers will be re-queued until it has been installed. (see Installing a Provider section above)